java/all/src/main/scala/org/typelevel/otel4s/java/OtelJava.scala

```diff
@@ -68,6 +68,8 @@ object OtelJava {
       def propagators: ContextPropagators[F] = contextPropagators
       def meterProvider: MeterProvider[F] = metrics.meterProvider
       def tracerProvider: TracerProvider[F] = traces.tracerProvider
+
+      override def toString: String = jOtel.toString()
```

Comment:
I wonder if the Java `OpenTelemetry#toString` is not guaranteed to be nice in general, but we can still use `propgators.toString`, `meterProvider.toString`, `tracerProvider.toString` etc. to put together our own `toString`?

Reply:
The uninformative `.toString` you're getting is the default from `java.lang.Object`.

In general, the Java Otel API is built on interfaces.  I wrote 99% of my Java before interfaces could have default implementations, and I don't know whether an interface can override `java.lang.Object#toString`.  Legal or not, I can confirm some, including `OpenTelemetry`, don't.  "Not guaranteed to be nice" is correct.

Some _implementations_ have a `.toString`, like `DefaultContextPropagators`.  Some do not, like `DefaultTracerProvider`.  So you'd make some progress going down that road, but won't be rid of the hexadecimal hashCodes, and ... there can be multiple implementations of these interfaces.  See "no guarantees."

The [golden rule of software quality](https://www.haskellforall.com/2020/07/the-golden-rule-of-software-quality.html) tells us to upstream better `.toString`s to Java, but that's not what @sherriesyt volunteered for.  If that's interesting anyway, I can help.  If it's not... I've been writing Java for a quarter century, and another hour probably wouldn't kill me.

Reply:
I don't mind dabbling in Java (though I'm rusty) -- can you point me to where to start? Is this somewhere in [opentelemetry-java](https://github.com/open-telemetry/opentelemetry-java/tree/main/api/all)?

Reply:
Yes.  The [`noop`](https://github.com/open-telemetry/opentelemetry-java/blob/93bcdf28706c1b6830394dc3a6d4dae1844e1c90/api/all/src/main/java/io/opentelemetry/api/OpenTelemetry.java#L31) value in your test is [implemented here](https://github.com/open-telemetry/opentelemetry-java/blob/93bcdf28706c1b6830394dc3a6d4dae1844e1c90/api/all/src/main/java/io/opentelemetry/api/DefaultOpenTelemetry.java), and a `.toString` in that class would make it nicer.

The one I would expect to see in production is [here](https://github.com/open-telemetry/opentelemetry-java/blob/93bcdf28706c1b6830394dc3a6d4dae1844e1c90/sdk/all/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java#L123), and already has a thorough `toString`.  So while the easy one to test is in rough shape, the good news is the common production one should already work pretty well!

If you decide to chase this upstream, the build tool is Gradle.  I've used `gradle publishToMavenLocal` before.  You can see the version number it generates in the output (I don't have it handy to be exact), and you can put that version into build.sbt to see how your work interacts with Otel4s.